rfc/20240529-OpenTofu-Specific-Code-Override.md

```diff
@@ -0,0 +1,160 @@
+
+TODO: run each user-story with the existing POC, and try to understand if we missed something that this solution is not addressing.
+
+### Language Server and IDE plugins
+We want to give our users the best possible experience when they work with OpenTofu in their IDE. For a comfortable experience our users need features like code completion, syntax highlighting and validation, jump to definition, and others when they work with OpenTofu. If we add support for `.tofu` files, we need to make sure that the IDE plugins that support OpenTofu will also support the new file extensions.
+
+#### VSCode
+Today, [VSCode](https://code.visualstudio.com/) supports Terraform using [Terraform Language Server](https://github.com/hashicorp/terraform-ls) and [Terraform Extension](https://marketplace.visualstudio.com/items?itemName=HashiCorp.terraform). The VSCode Extension is using the Terraform Language Server as part of its implementation.
+Both projects were already forked by a member of the community - [OpenTofu Language Server](https://github.com/gamunu/opentofu-ls) and [OpenTofu VSCode Extension](https://github.com/gamunu/vscode-opentofu). By looking at the projects, it looks like some work has been done, but support is still lacking for features that were introduced as part of OpenTofu 1.7. We should talk to the maintainer and see how we can contribute and make sure the plugins are up-to-date for supporting the latest OpenTofu features. In addition, we should make sure that the plugins will support the new file extensions that we are going to introduce.
+
+#### JetBrains
+Currently, for Terraform, [Jetbrains IDEs](https://www.jetbrains.com/ides/) has the [Terraform and HCL plugin](https://plugins.jetbrains.com/plugin/7808-terraform-and-hcl) that is not using the LSP protocol. The license of the plugin is [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0), based on the Marketplace link mentioned before, although I didn't find a clear license in the [source code repository](https://github.com/JetBrains/intellij-plugins/tree/f99102c5d7f762f3d86a547b69da17bedb0eaf93/terraform).
+We should talk to the maintainer. We might be able to contribute to the plugin to support OpenTofu, or more likely, create a new plugin that supports only OpenTofu (and the new file extensions suggested in this RFC).
+In case we want to create a new plugin, we might want to consider that on 2023, Jetbrains started [supporting LSP for Plugin Developers](https://blog.jetbrains.com/platform/2023/07/lsp-for-plugin-developers/) and we can use the [OpenTofu Language Server](https://github.com/gamunu/opentofu-ls) for the development of the new plugin. Although, it can only support paid IntelliJ-based IDEs, as explained [here](https://blog.jetbrains.com/platform/2023/07/lsp-for-plugin-developers/).
+
+#### Configurable plugin to support `.tf` and `.tofu` files based on user preference
```

Comment:
How do these plugins deal with a "collision", having two such plugins run on the same file?

Reply:
Good question. No idea. It's up to each IDE's implementation. We'll have to figure it out.